### ISSUE TITLE
图像编辑的时候偶现报错

### ISSUE THREAD

**Issue by SongDI911** [ByteDance-Seed/Bagel#7]
local variable 'cfg_img_v_t' referenced before assignment

**Comment by SongDI911** [comment]
import gradio as gr
import os
from copy import deepcopy
from typing import (
    Any,
    AsyncIterable,
    Callable,
    Dict,
    Generator,
    List,
    NamedTuple,
    Optional,
    Tuple,
    Union,
)
import requests
from io import BytesIO

from PIL import Image
import torch
from accelerate import infer_auto_device_map, load_checkpoint_and_dispatch, init_empty_weights

from data.transforms import ImageTransform
from data.data_utils import pil_img2rgb, add_special_tokens
from modeling.bagel import (
    BagelConfig, Bagel, Qwen2Config, Qwen2ForCausalLM, SiglipVisionConfig, SiglipVisionModel
)
from modeling.qwen2 import Qwen2Tokenizer
from modeling.bagel.qwen2_navit import NaiveCache
from modeling.autoencoder import load_ae
from safetensors.torch import load_file

# 模型初始化 (单例)
model_path = "/home/ubuntu/gxs/LLaMA_MODELS/BAGEL-7B-MoT"

# LLM config preparing
llm_config = Qwen2Config.from_json_file(os.path.join(model_path, "llm_config.json"))
llm_config.qk_norm = True
llm_config.tie_word_embeddings = False
llm_config.layer_module = "Qwen2MoTDecoderLayer"

# ViT config preparing
vit_config = SiglipVisionConfig.from_json_file(os.path.join(model_path, "vit_config.json"))
vit_config.rope = False
vit_config.num_hidden_layers = vit_config.num_hidden_layers - 1

# VAE loading
vae_model, vae_config = load_ae(local_path=os.path.join(model_path, "ae.safetensors"))

# Bagel config preparing
config = BagelConfig(
    visual_gen=True,
    visual_und=True,
    llm_config=llm_config, 
    vit_config=vit_config,
    vae_config=vae_config,
    vit_max_num_patch_per_side=70,
    connector_act='gelu_pytorch_tanh',
    latent_patch_size=2,
    max_latent_size=64,
)

with init_empty_weights():
    language_model = Qwen2ForCausalLM(llm_config)
    vit_model      = SiglipVisionModel(vit_config)
    model          = Bagel(language_model, vit_model, config)
    model.vit_model.vision_model.embeddings.convert_conv2d_to_linear(vit_config, meta=True)

# Tokenizer Preparing
tokenizer = Qwen2Tokenizer.from_pretrained(model_path)
tokenizer, new_token_ids, _ = add_special_tokens(tokenizer)

# Image Transform Preparing
vae_transform = ImageTransform(1024, 512, 16)
vit_transform = ImageTransform(980, 224, 14)

max_mem_per_gpu = "40GiB"  # Modify it according to your GPU setting

device_map = infer_auto_device_map(
    model,
    max_memory={i: max_mem_per_gpu for i in range(torch.cuda.device_count())},
    no_split_module_classes=["Bagel", "Qwen2MoTDecoderLayer"],
)
print(device_map)

same_device_modules = [
    'language_model.model.embed_tokens',
    'time_embedder',
    'latent_pos_embed',
    'vae2llm',
    'llm2vae',
    'connector',
    'vit_pos_embed'
]

for k in same_device_modules:
    device_map[k] = device_map[same_device_modules[0]]

model = load_checkpoint_and_dispatch(
    model,
    checkpoint=os.path.join(model_path, "ema.safetensors"),
    device_map=device_map,
    offload_buffers=True,
    dtype=torch.bfloat16,
)

model = model.eval()
print('Model loaded')

from inferencer import InterleaveInferencer

inferencer = InterleaveInferencer(
    model=model, 
    vae_model=vae_model, 
    tokenizer=tokenizer, 
    vae_transform=vae_transform, 
    vit_transform=vit_transform, 
    new_token_ids=new_token_ids
)


def run_inference(
    prompt: str,
    image: gr.File = None,
    mode: str = "生成",
    cfg_text: float = 4.0,
    cfg_img: float = 1.0,
    timesteps: int = 50,
    think: bool = False,
    timestep_shift: float = 3.0,
):
    hyper_params = {
        "cfg_text_scale": cfg_text,
        "cfg_img_scale": cfg_img,
        "num_timesteps": timesteps,
        "cfg_interval": [0.4, 1.0],
        "timestep_shift": timestep_shift,
        "cfg_renorm_min": 1.0,
        "cfg_renorm_type": "global",
    }
    
    try:
        max_think_token_n = 1000
        if mode == "编辑" and not image:
            raise ValueError("编辑模式需要上传图片")
            
        inputs = {"text": prompt, "think": think}
        if think:
            hyper_params["max_think_token_n"] = max_think_token_n
        if mode == "理解":
            output = inferencer(image=image, text=prompt, understanding_output=True, **hyper_params)
        elif mode == "编辑":
            hyper_params = {
                "cfg_text_scale": cfg_text,
                "cfg_img_scale": cfg_img,
                "num_timesteps": timesteps,
                "cfg_interval": [0.4, 1.0],
                "timestep_shift": timestep_shift,
                "cfg_renorm_min": 1.0,
                "max_think_token_n": max_think_token_n,
                "do_sample": False,
                "cfg_renorm_type": "text_channel",
            }
            output = inferencer(image=image, text=prompt, think=think, **hyper_params)
        else:
            output = inferencer(text=prompt, think=think, **hyper_params)

        print(f"think: {think}, out_text: {output.get('text', '')}")
        return output['image'], output.get('text', '')
    except Exception as e:
        return None, f"错误: {str(e)}"

with gr.Blocks(title="BAGEL 图像生成器") as app:
    gr.Markdown("# 🎨 BAGEL-7B 多模态交互系统")
    
    with gr.Tabs():
        with gr.Tab("文本生成图像"):
            with gr.Row():
                with gr.Column():
                    text_input = gr.Textbox(label="描述词", placeholder="输入详细描述...")
                    cfg_text = gr.Slider(1.0, 10.0, value=4.0, label="文本引导强度")
                    cfg_img = gr.Slider(0.5, 5.0, value=1.0, label="图像生成强度")
                    steps = gr.Slider(10, 100, value=50, step=1, label="生成步数")
                    timestep_shift = gr.Slider(0.0, 10.0, value=3.0, label="时间步长偏移")
                    think_check = gr.Checkbox(label="think模式", value=False)
                    
                    generate_btn = gr.Button("生成", variant="primary")
                
                output_image = gr.Image(label="生成结果", type="pil")
                output_text = gr.Textbox(label="推理过程")

            generate_btn.click(
                fn=run_inference,
                inputs=[text_input, gr.State(None), gr.State("生成"), cfg_text, cfg_img, steps, think_check, timestep_shift],
                outputs=[output_image, output_text]
            )
        
        with gr.Tab("图像编辑"):
            with gr.Row():
                with gr.Column():
                    edit_image = gr.Image(label="上传原图", type="pil")
                    edit_prompt = gr.Textbox(label="编辑指令", placeholder="输入编辑描述...")
                    edit_cfg = gr.Slider(2.0, 8.0, value=4.0, label="编辑强度")
                    edit_cfg_img = gr.Slider(0.5, 5.0, value=1.0, label="图像生成强度")
                    edit_timestep_shift = gr.Slider(0.0, 10.0, value=3.0, label="时间步长偏移")
                    edit_think_check = gr.Checkbox(label="think模式", value=False)
                    edit_btn = gr.Button("执行编辑", variant="primary")
                
                edit_output = gr.Image(label="编辑结果", type="pil")
                edit_output_text = gr.Textbox(label="过程")

            edit_btn.click(
                fn=run_inference,
                inputs=[edit_prompt, edit_image, gr.State("编辑"), edit_cfg, edit_cfg_img, gr.State(50), edit_think_check, edit_timestep_shift],
                outputs=[edit_output, edit_output_text]
            )
            
            gr.Examples(
                examples=[["test_images/women.jpg", "换上赛博朋克风格的服装"],
                          ["test_images/octupusy.jpg", "转化为水彩画风格"]],
                inputs=[edit_image, edit_prompt]
            )

        with gr.Tab("图像理解"):
            with gr.Row():
                understand_image = gr.Image(label="上传图片", type="pil")
                understand_text_input = gr.Textbox(label="描述词", placeholder="输入详细描述...")
                understand_btn = gr.Button("分析内容", variant="primary")
            understand_output = gr.Textbox(label="分析结果")
            
            understand_btn.click(
                fn=run_inference,
                inputs=[understand_text_input, understand_image, gr.State("理解")],
                outputs=[gr.Image(visible=False), understand_output]
            )

if __name__ == "__main__":
    app.launch(server_name="0.0.0.0", server_port=7868)

**Comment by bladewaltz1** [comment]
thanks. It's a bug. We'll fix it today.

**Comment by SongDI911** [comment]
当cfg_img_scale小于等于1，cfg_renorm_type=text_channe ，cfg_text_scale>1l时 必现
https://github.com/ByteDance-Seed/Bagel/blob/bfb33dc884dc0dc9abfe3512667f7c9e29208b4c/modeling/bagel/bagel.py#L824C12-L824C26

**Comment by Andy1621** [comment]
Yes. Since we found that the `cfg_img_scale` parameter aids in image editing, we've set its default value to be greater than 1.0. I will fix the bug with `cfg_img_scale=1.0` later.

**Comment by Andy1621** [comment]
@SongDI911 I have update the code, please check https://github.com/ByteDance-Seed/Bagel/blob/e556da202496d173990b4aac9c894b71e04a47bd/modeling/bagel/bagel.py#L831-L834

**Comment by SongDI911** [comment]
> [@SongDI911](https://github.com/SongDI911) I have update the code, please check
> 
> [Bagel/modeling/bagel/bagel.py](https://github.com/ByteDance-Seed/Bagel/blob/e556da202496d173990b4aac9c894b71e04a47bd/modeling/bagel/bagel.py#L831-L834)
> 
> Lines 831 to 834 in [e556da2](/ByteDance-Seed/Bagel/commit/e556da202496d173990b4aac9c894b71e04a47bd)
> 
>  if cfg_img_scale > 1.0: 
>      v_t = cfg_img_v_t + cfg_img_scale * (v_t_text - cfg_img_v_t) 
>  else: 
>      v_t = v_t_text

Thanks
